### ISSUE TITLE
Allow running tests in parallel

### ISSUE THREAD

**Issue by orf** [Homebrew/homebrew-test-bot#596]
Hello :wave:

Some core formula like sqlite (https://github.com/Homebrew/homebrew-core/pull/74485) hit a recurring issue where builds take _forever_. Here's one that's been going [for over 9 hours](https://github.com/Homebrew/homebrew-core/actions/runs/722540993). It seems like it would be trivial to parallelize the `brew test-bot` command, especially as we are using Github Actions, and reduce the time it takes to check these changes in. For context, SQLite 3.35.3 has a nasty bug that is fixed in 3.35.4 and it would be nice to be able to release this without running into CI timeouts.

Github Actions allows you to create dynamic matrices based on prior steps. For example:

```yaml
jobs:
    whatever:
        steps:
            run: echo "::set-output name=matrix::$(some-command)"
        outputs:
            matrix: ${{ steps.set-matrix.outputs.matrix }}

    run_something:
         needs: whatever
         strategy:
             matrix:
                 something: ${{ fromJson(needs.whatever.outputs.matrix) }}
         steps:
            ....
```

In the above example we have a step that executes `some-command`, and uses the JSON output from this command to feed into the `run_something` matrix.

I propose that we adapt the `brew test-bot` command to support a dynamic matrix. We can do this quite simply: We add a `--matrix` flag of some kind that outputs a JSON array of shards. If `sqlite3` tests needs to build 1,000 packages, and we specify that each job should handle 100 packages then this would be a simple array of `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`. If we're building a package that needs to build only 200 packages to test, then it might just be `[1, 2]`. As this only really makes a difference on some core packages we don't want to always launch a matrix of 10 jobs if each one is only going to run 3 or 4 tests. So there's some smartness that could be added to account for this which is different from just `idx % total`.

The downstream test would depend on this output and would use the value to shard tests. So, given the above 1,000 packages that need to be tested across 10 different runners then we'd run `brew test-bot ${{ matrix.shard }} ${{ matrix.total_shards }}`, which would run 10% of the packages it needs to.

In theory this would result in a lot faster builds for packages that are heavily depended on. We could also run the `brew test-bot --matrix` command on a Ubuntu runner to save resources.

I'm not that familiar with Homebrew or the intricacies of how it's tested, so I hope this makes sense!

**Comment by Bo98** [comment]
Yes, there is a goal to improve this. There's some groundwork being put into place at: https://github.com/Homebrew/homebrew-test-bot/pull/582, which would at least allow partial CI restarts. Futher iterations can allow parallelising tasks to multiple runners by splitting the dependent tree into subtrees.

> across 10 different runners

We are limited in resources so we probably won't be able to go that far. We currently have 4 runners per OS.

---

I'm also thinking about potentially applying some version rules here so that, for example, we don't spend an unnecessary amount of time testing a critical OpenSSL security release that is guaranteed to be ABI compatible. This would help with the "need this critical bugfix ASAP" case:

> SQLite 3.35.3 has a nasty bug that is fixed in 3.35.4

**Comment by carlocab** [comment]
We can't really run 10 parallel builds at a time, since we use self-hosted runners to build. We need the self-hosted runners since GitHub only reliably provides runners for macOS 10.15 when we also need to build for 10.14 and 11 (both ARM and Intel).

**Comment by orf** [comment]
I mean we can substitute 10 for "any number greater than 1" with the same effect.

>  GitHub only reliably provides runners for macOS 10.15 

Can you elaborate on "reliably"?

**Comment by carlocab** [comment]
Any splitting at all isn't quite supported in `test-bot` yet, see Bo's comment above.

GitHub provides runners for macOS 10.15 and (Intel) macOS 11, but that's [in private preview](https://github.com/actions/virtual-environments/issues/2486). We recently (ish, maybe it was a few months ago) switched all of our workflows that ran on (GitHub-provided) `macos-11` runners to `macos-latest` (i.e. 10.15) because they would spend ages waiting in queues and fail randomly because they were never assigned a VM.

**Comment by orf** [comment]
Ok I see. The 10.15 runners are still hosted by Github, but the impact of running parallel on only them is limited because we'd still be running sequential builds on the other self-hosted runners?

It seems like there's a lot to work on here, the general idea behind the ticket was to suggest auto-generating some kind of matrix using `test-bot`. If that's not going to work then we can close this 👍.

**Comment by carlocab** [comment]
Yea, it's not quite that simple. Generating the matrix correctly is quite complicated (since we have to worry about the dependency graph), and even if we had that matrix, we still need to teach `test-bot` what to do with that matrix.

**Comment by MikeMcQuaid** [comment]
> It seems like it would be trivial to parallelize the `brew test-bot` command
> We can do this quite simply:

@orf If things are trivial/simple then please open pull requests instead of issues. Relatedly, if things are trivial/simple, would produce huge speedups and they aren't already done: chances are they aren't trivial/simple. I know it's not your intention but this assumption is a little insulting.

> we don't spend an unnecessary amount of time testing a critical OpenSSL security release that is guaranteed to be ABI compatible. This would help with the "need this critical bugfix ASAP" case:

@Bo98 I'd suggest we verify the ABI compatibility ourselves somehow. I'm not sure how it is possible but I know it is in some form (because I remember when I worked on KDE it'd break the build if you broke the API on a patch release).
